service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java

```diff
@@ -164,6 +179,109 @@ private void processCredentialsMessageGetOperation(final Message<JsonObject> reg
         getCredentials(tenantId, type, authId, result -> reply(regMsg, result));
     }
 
+    private void processCredentialsMessageAddOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        if (!isValidCredentialsObject(payload)) {
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        addCredentials(tenantId, payload, result -> reply(regMsg, result));
+    }
+    
+    private void processCredentialsMessageUpdateOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        if (!isValidCredentialsObject(payload)) {
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        updateCredentials(tenantId, payload, result -> reply(regMsg, result)); 
+    }
+    
+    private void processCredentialsMessageRemoveOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        final String deviceId = payload.getString(FIELD_DEVICE_ID);
+        if (deviceId == null) {
+            log.debug("credentials remove request did not contain device-id in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        
+        final String type = payload.getString(FIELD_TYPE);
+        if (type == null) {
+            log.debug("credentials remove request did not contain type in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+
+        final String authId = payload.getString(FIELD_AUTH_ID);
+        
+        removeCredentials(tenantId, deviceId, type, authId, result -> reply(regMsg, result));
+    }
+
+    boolean isValidCredentialsObject(final JsonObject credentials) {
```

Comment:
See my above comment on the previous commit regarding package visibility only for unit testing.
Can you please rewrite the test that it only relies on the public method and make these methods private access?
Thx!